<a href="https://colab.research.google.com/github/nceder/qpb4e/blob/main/code/Chapter%2006/Chapter_06.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# 6 Strings

# 6.1 Strings as sequences of characters

In [1]:
x = "Hello"
x[0]

'H'

In [2]:
x[-1]

'o'

In [3]:
x[1:]

'ello'

In [4]:
x = "Goodbye\n"
x = x[:-1]
x

'Goodbye'

In [5]:
len("Goodbye")

7

# 6.2 Basic string operations

In [6]:
x = "Hello " + "World"
x

'Hello World'

In [7]:
8 * "x"

'xxxxxxxx'

# 6.3 Special characters and escape sequences

## 6.3.2 Numeric (octal and hexadecimal) and Unicode escape sequences

In [8]:
'm'

'm'

In [9]:
'\155'

'm'

In [10]:
'\x6D'

'm'

In [11]:
'\n'

'\n'

In [12]:
'\012'

'\n'

In [13]:
'\x0A'

'\n'

In [14]:
unicode_a ='\N{LATIN SMALL LETTER A}'           #A
unicode_a

'a'

In [15]:
unicode_a_with_acute = '\N{LATIN SMALL LETTER A WITH ACUTE}'
unicode_a_with_acute

'á'

In [16]:
"\u00E1"                    #B

'á'

Here is how you can find the Unicode numbers (code points) for emojis and other complex characters in Python.

There are two main ways to do this: looking them up by **name** or inspecting a character you **already have**.

### 1. The "Name" Method (Best for finding new things)

If you know what you are looking for (like "poop"), but don't know the symbol or code, you can use the `unicodedata` module or the `\N{...}` escape sequence.

#### Using `\N` (The Easiest Way)

Python allows you to type characters by their official Unicode name directly in a string.

```python
# You can use the official name inside \N{...}
print("\N{PILE OF POO}") 
# Output: 💩

print("\N{CAT FACE}")
# Output: 🐱

```

#### Using `unicodedata` to search

If you want to search programmatically, you can use the `unicodedata` library.

```python
import unicodedata

# Get the character from the name
char = unicodedata.lookup("PILE OF POO")
print(char)  # Output: 💩

# Get the name from the character
name = unicodedata.name("💩")
print(name)  # Output: PILE OF POO

```

---

### 2. The "Inspector" Method (Getting the number)

If you have the emoji (you can copy/paste it into your code) and want to know its Python code point, use `ord()` combined with string formatting.

```python
emoji = "💩"

# 1. Get the integer value
code_point = ord(emoji)
print(code_point)  
# Output: 128169 (This is the decimal number)

# 2. Get the Hex representation (standard Unicode format)
hex_code = f"{code_point:X}" 
print(hex_code)
# Output: 1F4A9

```

### Important: `\u` vs `\U`

This is where most people get stuck with emojis.

* **`\u` (Little u):** Only works for 4-digit hex codes (older characters like `\u00E1` for 'á').
* **`\U` (Big U):** Must be used for emojis because they have larger numbers (5-6 digits). You must pad it with zeros to make it exactly **8 digits**.

So, for the poop emoji (`1F4A9`), you must write it like this:

```python
# WRONG (Too short for \u, not padded for \U)
# print("\u1F4A9") 

# CORRECT (Capital U, padded to 8 digits)
print("\U0001F4A9")
# Output: 💩

```

### Summary Table

| Character | Name | Decimal (`ord`) | Hex | Python String |
| --- | --- | --- | --- | --- |
| á | LATIN SMALL LETTER A WITH ACUTE | 225 | E1 | `"\u00E1"` |
| 💩 | PILE OF POO | 128169 | 1F4A9 | `"\U0001F4A9"` |
| 🐍 | SNAKE | 128013 | 1F40D | `"\U0001F40D"` |

Would you like a small script that prints out a range of emojis so you can see what neighbors the poop emoji has?

In [16]:
import unicodedata

# Get the character from the name
char = unicodedata.lookup("PILE OF POO")
print(char)  # Output: 💩

# Get the name from the character
name = unicodedata.name("💩")
print(name)  # Output: PILE OF POO

💩
PILE OF POO


In [ ]:
print("\U0001F340", )

🍀


In [19]:
import unicodedata

dir(unicodedata.name)

['__call__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__name__',
 '__ne__',
 '__new__',
 '__qualname__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__self__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__text_signature__']

## 6.3.3 Printing vs. evaluating strings with special characters

In [20]:
'a\n\tb'

'a\n\tb'

In [23]:
print('a\n\tb')

a
	b


In [24]:
print("abc\n")

abc



In [25]:
print("abc\n", end="")

abc


# 6.4 String methods

## 6.4.1 The split and join string methods


In [30]:
" ".join(["join", "puts", "spaces", "between", "elements"])

'join puts spaces between elements'

In [31]:
"::".join(["Separated", "with", "colons"])

'Separated::with::colons'

In [32]:
"".join(["Separated", "by", "nothing"])

'Separatedbynothing'

In [34]:
x = "You\t\t can have tabs\t\n \t and newlines \n\n " \
               "mixed in"
x.split()

['You', 'can', 'have', 'tabs', 'and', 'newlines', 'mixed', 'in']

In [35]:
x = "Mississippi"
x.split("ss")

['Mi', 'i', 'ippi']

In [36]:
x = 'a b c d'
x.split(' ', 1)

['a', 'b c d']

In [38]:
x.split(' ', 2)

['a', 'b', 'c d']

In [39]:
x.split(' ', 9)

['a', 'b', 'c', 'd']

In [40]:
x.split(None, 2)

['a', 'b', 'c d']

### Quick Check: split and join
How could you use split and join to change all the whitespace in string x to dashes, such as changing "this is a test" to "this-is-a-test"?


In [41]:
x = "this is a test"
print(x, "-".join(x.split()))

this is a test this-is-a-test


In [ ]:
x = "this is a test"
"-".join(x.split())

'this-is-a-test'

## 6.4.2 Converting strings to numbers

In [42]:
float('123.456')

123.456

In [43]:
float('xxyy')

ValueError: could not convert string to float: 'xxyy'

In [44]:
int('3333')

3333

In [45]:
int('123.456')

ValueError: invalid literal for int() with base 10: '123.456'

In [46]:
int('10000', 8)

4096

In [47]:
int('101', 2)

5

In [49]:
int('666', 36)

7998

In [50]:
6*(36**2) + 6*(36**1) + 6*(36**0)

7998

In [51]:
int('ff', 16)

255

In [52]:
int('123456', 6)

ValueError: invalid literal for int() with base 6: '123456'

That line of code is a translator. You are telling Python: **"Take the string `'ff'`, pretend it is written in Base 16 (Hexadecimal), and convert it into a normal Base 10 integer."**

Here is the breakdown of why it equals 255.

### 1. The Function Arguments

`int('ff', 16)` has two parts:

* **`'ff'`**: The text you want to convert.
* **`16`**: The "base" or language that text is written in.

### 2. The "Dictionary" (Hexadecimal)

In our normal counting (Base 10), we have digits 0–9.
In Hexadecimal (Base 16), we need more digits, so we use letters:

* `0-9` are the same.
* `a` = 10
* `b` = 11
* `c` = 12
* `d` = 13
* `e` = 14
* **`f` = 15**

### 3. The Math

Just like the number **21** in normal math means `(2 * 10) + (1 * 1)`, hexadecimal uses powers of **16**.

So for **`ff`**:

1. The right `f` is in the "ones" spot. It is worth **15**.
2. The left `f` is in the "sixteens" spot. It is worth **15 * 16**.

**The calculation:**


### Why this matters for your Unicode exploration

This function is actually the **reverse** of what you were doing with `\u` codes.

When you see a Unicode code point like the Poop Emoji (`1F4A9`), that is a Hexadecimal number. You can use `int()` to find out its real ID number:

```python
# Convert the Hex string "1F4A9" (Base 16) to an Integer
print(int('1F4A9', 16))
# Output: 128169

```

This confirms that the "ID number" for the poop emoji is 128,169.

Would you like to see how to do the opposite—turn the number `255` back into `'ff'`?

### Quick Check: Strings to Numbers
Which of the following will not be converted to numbers, and why?
```
int('a1')
int('12G', 16)
float("12345678901234567890")
int("12*2")
```

In [56]:
int("12*2")

ValueError: invalid literal for int() with base 10: '12*2'

In [57]:
float("12345678901234567890")

1.2345678901234567e+19

In [58]:
int('12G', 16)

ValueError: invalid literal for int() with base 16: '12G'

In [59]:
int('a1')

ValueError: invalid literal for int() with base 10: 'a1'

```python
results = {}

# Test 1: int('a1')
try:
    results['int("a1")'] = f"Success: {int('a1')}"
except Exception as e:
    results['int("a1")'] = f"Error: {type(e).__name__} - {e}"

# Test 2: int('12G', 16)
try:
    results["int('12G', 16)"] = f"Success: {int('12G', 16)}"
except Exception as e:
    results["int('12G', 16)"] = f"Error: {type(e).__name__} - {e}"

# Test 3: float("12345678901234567890")
try:
    results['float("12345678901234567890")'] = f"Success: {float('12345678901234567890')}"
except Exception as e:
    results['float("12345678901234567890")'] = f"Error: {type(e).__name__} - {e}"

# Test 4: int("12*2")
try:
    results['int("12*2")'] = f"Success: {int('12*2')}"
except Exception as e:
    results['int("12*2")'] = f"Error: {type(e).__name__} - {e}"

print(results)


```

```text
{'int("a1")': "Error: ValueError - invalid literal for int() with base 10: 'a1'", "int('12G', 16)": "Error: ValueError - invalid literal for int() with base 16: '12G'", 'float("12345678901234567890")': 'Success: 1.2345678901234567e+19', 'int("12*2")': "Error: ValueError - invalid literal for int() with base 10: '12*2'"}


```

Here is the breakdown. Technically, **three** of these will fail and raise a `ValueError`, and only **one** will successfully convert to a number.

### The Failures (and Why)

1. **`int('a1')`** → **FAILS**
* **Reason:** When you don't specify a base, Python assumes Base 10 (standard decimal). Base 10 only allows digits `0-9`. The letter `a` is not allowed.
* **Fix:** If you meant hex, you needed `int('a1', 16)`.


2. **`int('12G', 16)`** → **FAILS**
* **Reason:** You told Python to use Base 16 (Hexadecimal). Hexadecimal digits go from `0-9` and `A-F`. The letter **`G`** is the 17th letter, so it is "out of bounds" for Base 16.
* **Fix:** Base 16 stops at F. You would need Base 17 or higher to handle G (`int('12G', 17)` actually works!).


3. **`int("12*2")`** → **FAILS**
* **Reason:** `int()` converts **strings to numbers**, it does not solve math problems. It sees `*` as an invalid character, just like a letter.
* **Fix:** If you want to solve math inside a string, you use `eval("12*2")`, though that can be dangerous if the string comes from an untrusted user.



### The Success

4. **`float("12345678901234567890")`** → **WORKS**
* **Result:** `1.2345678901234567e+19`
* **Reason:** Python creates a floating-point number. Note that it converts it to scientific notation (`e+19`) because floats lose precision with numbers this big (it can't remember every single digit perfectly), but it is still a valid number.

In [ ]:
# G -> 16
# 12G -> (base 17) = 1*17**2 + 2*17*1 + 16
print(1*17**2 + 2*17**1 + 16*(17**0))
print(int('12g', 17))

339
339


## 6.4.3 Getting rid of extra whitespace

In [62]:
x = "  Hello,    World\t\t "
x.strip()

'Hello,    World'

In [63]:
x.lstrip()

'Hello,    World\t\t '

In [64]:
x.rstrip()

'  Hello,    World'

In [66]:
import string
string.whitespace

' \t\n\r\x0b\x0c'

In [67]:
" \t\n\r\v\f"

' \t\n\r\x0b\x0c'

In [68]:
x = "www.python.org"
x.strip("w")

'.python.org'

In [69]:
x.strip("gor")

'www.python.'

In [70]:
x.strip(".gorw")

'python'

In [71]:
x = "www.python.org"
x.removeprefix("www.")

'python.org'

In [72]:
x.removesuffix(".org")

'www.python'

### Quick Check: strip
If the string x equals "(name, date),\n", which of the following returns a string containing "name, date" ?
```
1. x.rstrip("),")
2. x.strip("),\n")
3. x.strip("\n)(,")
4. x.strip("\n)(,") will remove the newline as well as the comma and parentheses.
```

In [73]:
x = "(name,date),\n"
x.strip("\n)(,")

'name,date'

## 6.4.4 String searching

In [74]:
x = "Mississippi"
x.find("ss")

2

In [75]:
x.find("zz")

-1

In [76]:
x = "Mississippi"
x.find("ss", 3)

5

In [77]:
x.find("ss", 0, 3)

-1

In [78]:
x = "Mississippi"
x.rfind("ss")

5

In [79]:
x = "Mississippi"
x.count("ss")

2

In [80]:
x = "fsssf"
x.count('ss')

1

In [81]:
x = "Mississippi"
x.startswith("Miss")

True

In [82]:
x.startswith("Mist")

False

In [83]:
x.endswith("pi")

True

In [84]:
x.endswith("p")

False

In [85]:
x.endswith(("i", "u"))

True

In [86]:
x.endswith(("xyi", "i"))

True

## 6.4.5 Modifying strings

In [87]:
x = "Mississippi"
x.replace("ss", "+++")

'Mi+++i+++ippi'

In [88]:
x = "~x ^ (y % z)"
table = x.maketrans("~^()", "!&[]")
x.translate(table)

'!x & [y % z]'

In [89]:
x = "~x ^ (y % z)"
table = x.maketrans("", "", "()~^%")
x.translate(table)

'x  y  z'

### Quick Check: String searching
If you want to see whether a line ends with the string "rejected", what string method would you use? Are there any other ways you could get the same result?


In [90]:
y = "I hope you are doing well and don't feel rejected"
y.endswith("rejected")


True

In [91]:
y[-len("rejected"):] == 'rejected'

True

In [93]:
y.removesuffix("di")

"I hope you are doing well and don't feel rejected"

## 6.4.6 Modifying strings with list manipulations

In [94]:
text = "Hello, World"
wordList = list(text)
wordList[6:] = []
wordList.reverse()
text = "".join(wordList)
print(text)

,olleH


### Quick Check: Modifying strings
What would be a quick way to change all punctuation in a string to spaces?


In [97]:
puncts = string.punctuation
print(puncts)
punct = str.maketrans(puncts, " "*len(puncts))
x = "This is text, with: punctuation! Right?"
x.translate(punct)


!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


'This is text  with  punctuation  Right '

In [98]:
punct = str.maketrans("!.,:;-?", "       ")
x = "This is text, with: punctuation! Right?"
x.translate(punct)

'This is text  with  punctuation  Right '

# 6.4.7 Useful methods and constants

In [99]:
x = "123"
x.isdigit()

True

In [100]:
x.isalpha()

False

In [101]:
x = "M"
x.islower()

False

In [102]:
x.isupper()

True

### Try This: String operations
Suppose that you have a list of strings in which some (but not necessarily all) of the strings begin and end with the double quote character:
```
x = ['"abc"', 'def', '"ghi"', '"klm"', 'nop']
```
What code would you use on each element to remove just the double quotes?


In [104]:
x = ['"abc"', 'def', '"ghi"', '"klm"', 'nop']
x = [s.strip('"') for s in x]
print(x)

['abc', 'def', 'ghi', 'klm', 'nop']


In [105]:
x = ['"abc"', 'def', '"ghi"', '"klm"', 'nop']
for item in x:
    print(item.strip('"'))

abc
def
ghi
klm
nop


# 6.5 Converting from objects to strings

In [106]:
repr([1, 2, 3])

'[1, 2, 3]'

In [107]:
x = [1]
x.append(2)
x.append([3, 4])
'the list x is ' + repr(x)

'the list x is [1, 2, [3, 4]]'

In [108]:
repr(len)

'<built-in function len>'

# 6.6 Using the format method

## 6.6.1 The format method and positional parameters

In [109]:
"{0} is the {1} of {2}".format("Ambrosia", "food", "the gods")

'Ambrosia is the food of the gods'

In [110]:
"{{Ambrosia}} is the {0} of {1}".format("food", "the gods")

'{Ambrosia} is the food of the gods'

## 6.6.2 The format method and named parameters

In [111]:
"{food} is the food of {user}".format(food="Ambrosia",
      user="the gods")

'Ambrosia is the food of the gods'

In [112]:
"{0} is the food of {user[1]}".format("Ambrosia",
         user=["men", "the gods", "others"])


'Ambrosia is the food of the gods'

## 6.6.3 Format specifiers

In [115]:
"{0:30} is the food of gods".format("Ambrosia")

'Ambrosia                       is the food of gods'

In [116]:
"{0:{1}} is the food of gods".format("Ambrosia", 30)

'Ambrosia                       is the food of gods'

In [117]:
"{food:{width}} is the food of gods".format(food="Ambrosia", width=10)

'Ambrosia   is the food of gods'

In [118]:
"{0:>10} is the food of gods".format("Ambrosia")

'  Ambrosia is the food of gods'

In [120]:
"{0:*>10} is the food of gods".format("Ambrosia")

'**Ambrosia is the food of gods'

### Quick Check: the format() method
What will be in x when the following snippets of code are executed?:
```
x = "{1:{0}}".format(3, 4)
x = "{0:$>5}".format(3)
x = "{a:{b}}".format(a=1, b=5)
x = "{a:{b}}:{0:$>5}".format(3, 4, a=1, b=5, c=10)
```

In [121]:
x = "{1:{0}}".format(3, 4)
print(x)
x = "{0:$>5}".format(3)
print(x)
x = "{a:{b}}".format(a=1, b=5)
print(x)
x = "{a:{b}}:{0:$>5}".format(3, 4, a=1, b=5, c=10)
print(x)

  4
$$$$3
    1
    1:$$$$3


The reason this happens is that **numbers are right-aligned by default** in Python.

When you wrote `{a:{b}}`, you essentially asked Python to "print the number `1` inside a box of size `5`." Since you didn't tell Python *where* to put the number inside that box, it used its default rule for integers: push it to the right.

Here is the step-by-step breakdown of what Python did versus what you expected:

### 1. Resolve the Inner Variable

First, Python looks at `{b}`. Since `b=5`, the format string changes from:
`"{a:{b}}"`    `"{a:5}"`

### 2. Apply the Format

Now Python sees `{a:5}`.

* **Value:** `a` is `1` (an integer).
* **Width:** `5` spaces.
* **Alignment:** NOT SPECIFIED.

**This is the key moment:**

* If `a` were a **String** (text), Python defaults to **Left** alignment (`"1    "`).
* If `a` is a **Number** (int/float), Python defaults to **Right** alignment (`"    1"`).

Because `1` is a number, Python pushes it to the right side of the 5-character space.

### 3. The Result

* **Part 1 (`{a:{b}}`):** `"    1"` (Four spaces, then the number)
* **Part 2 (`:`):** `":"`
* **Part 3 (`{0:$>5}`):** `"$$$$3"`

**Total:** `"    1:$$$$3"`

---

### How to get the output you wanted

If you want the number to stay on the left (so the padding is on the right), you must explicitly add the `<` (left-align) symbol.

You can modify your format string to include the `<` inside the nested part:

```python
# Notice the '<' added before the inner bracket
x = "{a:<{b}}:{0:$>5}".format(3, 4, a=1, b=5)
print(x)

```

**Output:**

```text
1    :$$$$3

```

### Summary of Defaults

| Data Type | Default Alignment | Example `{val:5}` |
| --- | --- | --- |
| **Number** | Right (`>`) | `"    1"` |
| **String** | Left (`<`) | `"1    "` |

In [122]:
# Notice the '<' added before the inner bracket
x = "{a:<{b}}:{0:$>5}".format(3, 4, a=1, b=5)
print(x)

1    :$$$$3


# String interpolation with f-strings

In [124]:
value = 42
message = f"The answer is {value}"
print(message)

The answer is 42


In [125]:
primes = [1, 2, 3, 5]
f"sum of first 4 primes is {sum(primes)}"

'sum of first 4 primes is 11'

In [126]:
PI = 3.1415
print(f"PI is {PI:{10}.{2}}")

PI is        3.1


In [128]:
print(f"{PI=:{10}.{5}}")

PI=    3.1415


This is a great example of "nested" formatting in f-strings. Python is effectively doing this in two distinct steps: **constructing the rule** and then **applying the rule**.

Here is the breakdown of what is happening inside `{PI:{10}.{2}}`:

### Step 1: Build the "Format String"

First, Python looks at the format specifier (the part after the colon) and evaluates any expressions inside curly braces `{}`.

* `{10}` evaluates to the number `10`.
* `{2}` evaluates to the number `2`.
* The `.` is just plain text.

Python stitches these together to create the final format code: **`"10.2"`**.

So, your code is functionally equivalent to running this:

```python
# The f-string dynamically built the string "10.2"
print(format(PI, "10.2"))

```

### Step 2: Apply the "10.2" Rule

Now Python formats `PI` using the rule `"10.2"`. This rule has three hidden behaviors:

1. **Width (`10`):** The final string must be 10 characters long.
2. **Precision (`.2`):** Because you didn't specify a letter (like `f`), Python defaults to "General" (`g`) mode. In this mode, precision controls the **number of significant digits**, not decimal places.
* `3.1415` rounded to **2 significant digits** is `3.1`.


3. **Alignment (Default):** Since `PI` is a number, it defaults to **Right Alignment**.

### The Result

* The value `3.1` is 3 characters long.
* The target width is 10.
* Python adds 7 spaces to the left.

`"       3.1"`

### The "Missing Letter" Trap

If you expected the output to be `3.14` (two decimal places), you need to explicitly add the type code `f` (fixed point) to the end of your format string.

Compare these two:

```python
# Your current code (Implicit 'g' mode -> Significant Digits)
# Rule becomes "10.2"
print(f"PI is {PI:{10}.{2}}")  
# Output: "       3.1"

# With 'f' added (Fixed Point mode -> Decimal Places)
# Rule becomes "10.2f"
print(f"PI is {PI:{10}.{2}f}") 
# Output: "      3.14"

```

In [ ]:
# Your current code (Implicit 'g' mode -> Significant Digits)
# Rule becomes "10.2"

PI = 3.1415
print(f"PI is {PI:{10}.{2}}")
# Output: "       3.1"

# With 'f' added (Fixed Point mode -> Decimal Places)
# Rule becomes "10.2f"
print(f"PI is {PI:{30}.{24}f}")
# Output: "      3.14"

PI is        3.1
PI is     3.141500000000000181188398


What you are seeing is the "ghost in the machine." You peeled back the curtain too far and saw the messy reality of how computers store numbers.

**You ran into the limit of Floating Point Precision (IEEE 754).**

Here is the plain English explanation of "what the fuck" happened.

### 1. The "Secret Identity" of 3.1415

In the human world (Base 10), `3.1415` is a nice, clean, exact number.
In the computer world (Base 2 / Binary), `3.1415` is an **irrational, repeating fraction** that goes on forever. It's impossible to write it down perfectly in binary code.

It is exactly the same problem as trying to write **1/3** in decimal:

* **Math:** 
* **Paper:**  (forever)
* **Calculator:**  (it has to stop somewhere)

Your computer had to "cut off" the binary version of 3.1415 to fit it into memory (64 bits). The number it actually stored wasn't `3.1415`, it was this:

`3.1415000000000001811883976188288...`

### 2. You asked for "The Truth"

Usually, Python hides this ugly truth from you. When you print `3.1415`, Python thinks, *"They probably mean the clean version,"* and rounds it off for display.

But your code asked for **24 decimal places** (`.{24}f`).
This forced Python to bypass its "polite" rounding and show you exactly what was stored in the memory chips.

* **First 16 digits:** `3.141500000000000` (Accurate! The float worked.)
* **Next 8 digits:** `181188398` (Garbage! This is the "binary noise" at the edge of the precision.)

### 3. The Rule of Thumb

Standard Python floats (which are 64-bit "doubles") only have about **15 to 17 significant digits** of precision.

* **Safe Zone:** Digits 1–15 are usually trustworthy.
* **Danger Zone:** Any digit after the 16th is likely just a random artifact of the binary conversion.

### How to fix it

If you actually *need* that kind of extreme precision (e.g., for money or scientific satellites), do not use standard `float`. Use the `decimal` module, which works like human math.

```python
from decimal import Decimal

# Pass it as a STRING so Python doesn't turn it into a float first
PI = Decimal("3.1415") 

# Now request crazy precision
print(f"PI is {PI:{30}.{24}f}")

```

**Output:**

```text
PI is     3.141500000000000000000000

```

In [ ]:
from decimal import Decimal

# Pass it as a STRING so Python doesn't turn it into a float first
PI = Decimal("3.1415")

# Now request crazy precision
print(f"PI is {PI:{30}.{24}f}")

PI is     3.141500000000000000000000


In [135]:
primes = [1, 2, 3, 5]
f"sum of first 4 primes is {sum(primes)}"



'sum of first 4 primes is 11'

In [136]:
primes = [1, 2, 3, 5]
f"sum of first 4 primes is {sum(primes)=}"

'sum of first 4 primes is sum(primes)=11'

# 6.8 Formatting strings with %

In [137]:
"%s is the %s of %s" % ("Ambrosia", "food", "the gods")

'Ambrosia is the food of the gods'

In [138]:
"%s is the %s of %s" % ("Nectar", "drink", "gods")

'Nectar is the drink of gods'

In [139]:
"%s is the %s of the %s" % ("Brussels Sprouts", "food",
       "foolish")

'Brussels Sprouts is the food of the foolish'

In [140]:
x = [1, 2, "three"]
"The %s contains: %s" % ("list", x)

"The list contains: [1, 2, 'three']"

In [141]:
"Pi is <%-6.2f>" % 3.14159 # use of the formatting sequence: %–6.2f

'Pi is <3.14  >'

## 6.8.2 Named parameters and formatting sequences

In [142]:
"%(PI).2f"

'%(PI).2f'

In [143]:
num_dict = {'e': 2.718, 'PI': 3.14159}
print("%(PI).2f - %(PI).4f - %(e).2f" % num_dict)

3.14 - 3.1416 - 2.72


In [144]:
print("a")
print("b")

a
b


In [145]:
print("a", "b", "c")

a b c


In [146]:
print("a", "b", "c", sep="|")

a|b|c


In [149]:
print("a", "b", "c",sep="\U0001F595", end="\n\n")
print("a", "b", "c",sep="\U0001F595", end="\n\n", file=open("testfile.txt", "w"))

a🖕b🖕c



### Quick Check: Formatting strings with %
What would be in the variable x after the following snippets of code have executed?
```
x = "%.2f" % 1.1111
x = "%(a).2f" % {'a':1.1111}
x = "%(a).08f" % {'a':1.1111}
```

In [ ]:
x = "%.2f" % 1.1111
print(x)   # 1.11
x = "%(a).2f" % {'a':1.1111}
print(x) # 1.11
x = "%(a).08f" % {'a':1.1111}
print(x) #

1.11
1.11
1.11110000


That syntax is Python's older "printf-style" string formatting, and the part `%(a)` is a **Named Placeholder**.

It is basically doing a dictionary lookup inside the string.

Here is the breakdown of the code `"%(a).08f"`:

### 1. `%(a)`: The "Lookup"

This tells Python: *"Look at the dictionary I gave you on the right side, find the key `'a'`, and use its value here."*

* **Dictionary:** `{'a': 1.1111}`
* **Value Found:** `1.1111`

### 2. `.08f`: The "Rule"

This tells Python how to display that value.

* **`.`**: Start formatting the decimal part.
* **`08`**: Use **8 decimal places**. (Note: The leading zero here doesn't actually do anything special; `.08` is treated exactly the same as `.8`).
* **`f`**: Standard Fixed Point notation.

### 3. The Result

You asked for **8 decimal places**, but the number `1.1111` only has **4**.
Python obeys your command by adding "padding" zeros to the end to reach the required count.

`1.1111`  `1.11110000` (4 original digits + 4 padding zeros)

### Equivalent in Modern Python (f-strings)

Since you are using f-strings elsewhere, here is how you would write that exact same logic today. It is much easier to read:

```python
data = {'a': 1.1111}

# Old way (what you posted)
print("%(a).8f" % data)

# Modern way (f-string)
print(f"{data['a']:.8f}")

```

In [151]:
data = {'a': 1.1111}

# Old way (what you posted)
print("%(a).8f" % data)

# Modern way (f-string)
print(f"{data['a']:.8f}")

1.11110000
1.11110000


# 6.9 Bytes

A bytes object is similar to a string object but with an important difference: A string is an immutable sequence of Unicode characters, whereas a bytes object is a sequence of integers with values from 0 to 256. Bytes can be necessary when you’re dealing with binary data, such as reading from a binary data file.

The key thing to remember is that bytes objects may look like strings, but they can’t be used exactly like strings or combined with strings:

In [152]:
unicode_a_with_acute = '\N{LATIN SMALL LETTER A WITH ACUTE}'
unicode_a_with_acute

'á'

In [153]:
xb = unicode_a_with_acute.encode()   #A
xb

b'\xc3\xa1'

In [154]:
xb += 'A'

TypeError: can't concat str to bytes

In [155]:
xb.decode()

'á'

In [156]:
xb

b'\xc3\xa1'

In [157]:
len(xb)

2

### Quick Check: Bytes
For which of the following kinds of data would you want to use a string? For which could you use bytes?
1. Data file storing binary data
2. Text in a language with accented characters
3. Text with only uppercase and lowercase roman characters
4. A series of integers no larger than 255

1. Data file storing binary data

   Bytes. Because the data is binary, you're more concerned with the contents as numbers rather than text. Therefore, it would make sense to use bytes.
2. Text in a language with accented characters

   String. Python 3 strings are Unicode, so they can handle accented characters.
3. Text with only uppercase and lowercase roman characters

   String. Strings should be used for all text in Python 3.
4. A series of integers no larger than 255

   Bytes. A byte is an integer no larger than 255, so the bytes type is perfect for storing integers like this.


### Lab 6: Preprocessing Text
In processing raw text, it's quite often necessary to clean and normalize the text before doing anything else. If you want to find the frequency of words in text, for example, you can make the job easier if, before you start counting, you make sure that everything is lowercase (or uppercase, if you prefer) and that all punctuation has been removed. It can also make things easier to break the text into a series of words.

In this lab, the task is to read an excerpt of the first chapter of Moby Dick, make sure that everything is one case, remove all punctuation, and write the words one per line to a second file. Again, because I haven’t yet covered reading and writing files, the code for those operations is supplied below.
Your task is to come up with the code to replace the commented lines in the sample below:
```
with open("moby_01.txt") as infile, open("moby_01_clean.txt", "w") as outfile:
    for line in infile:
        # make all one case
        # remove punctuation
        # split into words
        # write all words for line
        outfile.write(cleaned_words)
```

### Run the cell below to get the file moby_01.txt

In [158]:
!wget https://raw.githubusercontent.com/nceder/qpb4e/main/code/Chapter%2006/moby_01.txt &> null  && echo Downloaded


Downloaded


In [171]:
# Create your solution in this cell

import string
import random
puncts = string.punctuation
punct = str.maketrans(puncts, " "*len(puncts))

# The Runic block starts at 16A0 (5792 in decimal)
start_index = int('16A0', 16)

# Print the first 24 runes (Elder Futhark)
elder_furthark_list = []
for i in range(24):
    char_code = start_index + i
    # We use chr() to convert an integer to its unicode character
    print(f"{chr(char_code)}  (Hex: {char_code:X})")
    elder_furthark_list.append(chr(char_code))





with open("moby_01.txt") as infile, open("moby_01_clean_david.txt", "w") as outfile:
    for line in infile:
        # make all one case
        line = line.lower()
        # remove punctuation
        cleaned_line = line.translate(punct)
        # split into words
        words = cleaned_line.split()
        # write all words for line
        for word in words:
            outfile.write(word + " " + random.choice(elder_furthark_list) + "\n")
        #cleaned_words = f" {random.choice(elder_furthark_list)}\n".join(words)
        #outfile.write(cleaned_words)



ᚠ  (Hex: 16A0)
ᚡ  (Hex: 16A1)
ᚢ  (Hex: 16A2)
ᚣ  (Hex: 16A3)
ᚤ  (Hex: 16A4)
ᚥ  (Hex: 16A5)
ᚦ  (Hex: 16A6)
ᚧ  (Hex: 16A7)
ᚨ  (Hex: 16A8)
ᚩ  (Hex: 16A9)
ᚪ  (Hex: 16AA)
ᚫ  (Hex: 16AB)
ᚬ  (Hex: 16AC)
ᚭ  (Hex: 16AD)
ᚮ  (Hex: 16AE)
ᚯ  (Hex: 16AF)
ᚰ  (Hex: 16B0)
ᚱ  (Hex: 16B1)
ᚲ  (Hex: 16B2)
ᚳ  (Hex: 16B3)
ᚴ  (Hex: 16B4)
ᚵ  (Hex: 16B5)
ᚶ  (Hex: 16B6)
ᚷ  (Hex: 16B7)


Your fix works - "precisely" (line 11) and "having" (line 12) are now on separate lines, and each word has a random Elder Futhark rune appended.

  Here's the markdown commentary you can add to your notebook:

  ## Comparison of AI-Generated Solutions for Lab 6

  This lab was solved using multiple approaches: my own solution, the book author's solution, and solutions generated by three different AI coding assistants (Google Colab/Gemini, GitHub Copilot, and Claude Code).

  ### Summary Table

  | Solution | Punctuation Removal | Word Output | Correct? | Lines of Code |
  |----------|---------------------|-------------|----------|---------------|
  | **David's** | `maketrans(punct, spaces)` | `for` loop + runes | Yes | 12 |
  | **Author's** | `maketrans('','',punct)` | `"\n".join()` | Yes | 10 |
  | **Colab/Gemini** | `for` loop + `replace()` | `' '.join()` | **No** | 10 |
  | **Copilot** | `maketrans('','',punct)` | `for` loop | Yes | 11 |
  | **Claude Code** | `maketrans('','',punct)` | `for` loop | Yes | 7 |

  ### Key Differences

  #### 1. Punctuation Removal Strategy

  **Two-argument `maketrans` (replace with spaces):**
  ```python
  punct = str.maketrans(string.punctuation, " "*len(string.punctuation))
  line.translate(punct)  # "hello, world" → "hello  world"
  Used by: David's solution

  Three-argument maketrans (delete entirely):
  punct = str.maketrans('', '', string.punctuation)
  line.translate(punct)  # "hello, world" → "hello world"
  Used by: Author, Copilot, Claude Code

  Loop with replace() (inefficient):
  for char in string.punctuation:
      line = line.replace(char, '')
  Used by: Colab/Gemini — This works but creates 32 intermediate strings (one per punctuation character), making it O(n*m) instead of O(n).

  2. Writing Words to File

  Loop approach (safest):
  for word in words:
      outfile.write(word + '\n')
  Used by: David, Copilot, Claude Code — Each word gets its own \n, impossible to miss line boundaries.

  Join approach (elegant but requires care):
  outfile.write("\n".join(words) + "\n")
  Used by: Author — Works correctly because of the final + "\n". Without it, words would concatenate across lines.

  Buggy join:
  outfile.write(' '.join(words) + '\n')
  Used by: Colab/Gemini — This is wrong! It writes all words from each input line on a single output line, space-separated, instead of one word per line.

  The Bug I Encountered

  My original solution used:
  cleaned_words = f" {rune}\n".join(words)
  outfile.write(cleaned_words)  # Missing final newline!

  The join() puts \n between words but not after the last word. Result: the last word of line N concatenated with the first word of line N+1 (e.g., "preciselyhaving").

  Fix: Either add + "\n" after join, or use a for loop.

  Lessons Learned

  1. AI assistants can produce buggy code — The Colab/Gemini solution has a fundamental bug that doesn't meet the requirements.
  2. str.maketrans has two forms — The 3-argument form ('', '', chars_to_delete) is cleaner when you want to remove characters entirely rather than replace them.
  3. join() doesn't add a trailing separator — This is a common source of bugs when processing files line-by-line.
  4. Always verify AI-generated code — Run it and check the output matches expectations!



In [ ]:
# Solution created by the author
import string

punct = str.maketrans("",  "", string.punctuation)

with open("moby_01.txt") as infile:
  with open("moby_01_clean.txt", "w") as outfile:
    for line in infile:
        # make all one case
        cleaned_line = line.lower()

        # remove punctuation
        cleaned_line = cleaned_line.translate(punct)

        # split into words
        words = cleaned_line.split()
        cleaned_words = "\n".join(words) + "\n"
        # write all words for line
        outfile.write(cleaned_words)


In [ ]:
# prompt: For every line the file moby_01.txt, make everything lower case,
#         remove punctuation, and split into words,
#         and write the words one per line  to the file moby_01_clean.txt

import string

# Open the input and output files
with open('moby_01.txt', 'r') as f_in, open('moby_01_clean.txt', 'w') as f_out:
  for line in f_in:
    # Convert the line to lowercase
    line = line.lower()

    # Remove punctuation
    for char in string.punctuation:
      line = line.replace(char, '')

    # Split the line into words
    words = line.split()

    # Write the words to the output file
    f_out.write(' '.join(words) + '\n')


In [ ]:
# Generated usinb CoPilot
# Prompt: or every line the file moby_01.txt, make everything lower case,
#         remove punctuation, and split into words.
#         Write the words one per line to the file moby_01_clean.txt.

import string

# Open the input file for reading
with open('moby_01.txt', 'r') as input_file:
    # Open the output file for writing
    with open('moby_01_clean.txt', 'w') as output_file:
        # Iterate over each line in the input file
        for line in input_file:
            # Convert the line to lowercase
            line = line.lower()

            # Remove punctuation from the line
            line = line.translate(str.maketrans('', '', string.punctuation))

            # Split the line into words
            words = line.split()

            # Write each word to the output file
            for word in words:
                output_file.write(word + '\n')